File: java/src/main/java/toolgood/algorithm/internals/functions/value/Function_PARAMETER.java
package toolgood.algorithm.internals.functions.value;

import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.Function_0;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_PARAMETER extends Function_0 {
    private final String name;

    public Function_PARAMETER(String name) {
        this.name = name;
    }

    @Override
    public String Name() {
        return "Parameter";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        String txt = name;
        if (tempParameter != null) {
            Operand r = tempParameter.apply(engine, txt);
            if (r != null) {
                return r;
            }
        }
        return engine.GetParameter(txt);
    }

    @Override
    public void toString(StringBuilder stringBuilder, boolean addBrackets) {
        stringBuilder.append(name);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NONE;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        ParameterType pt = new ParameterType();
        pt.Name = name;
        pt.Type = operandType;
        pt.Operator = op;
        pt.Value = val;
        result.add(pt);
    }
}

File: csharp/ToolGood.Algorithm/Internals/Functions/Value/Function_PARAMETER.cs
using System;
using System.Collections.Generic;
using System.Text;
using ToolGood.Algorithm.Enums;
using ToolGood.Algorithm.Internals;

namespace ToolGood.Algorithm.Internals.Functions.Value
{
	internal sealed class Function_Parameter : FunctionBase
	{
		private readonly string name;

		public Function_Parameter(string name)
		{
			this.name = name;
		}

		public override string Name => "Parameter";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var txt = name;
			if(tempParameter != null) {
				var r = tempParameter(engine, txt);
				if(r != null) return r;
			}
			return engine.GetParameter(txt);
		}
		public override void ToString(StringBuilder stringBuilder, bool addBrackets)
		{
			stringBuilder.Append(name);
		}
		public override OperandType GetResultType()
		{
			return OperandType.NONE;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			result.Add(new ParameterType() {
				Name = name,
				Type = operandType,
				Operator = op,
				Value = val
			});
		}
	}

}
